### PR TITLE
feat: ETLng Registry

### PR DESCRIPTION
**Pull request by godexsoft**
This PR introduces the `Registry`. A central component of the updated ETL. 
`RegistryInterface`, alongside declaring the interface of a Registry, documents what `hooks` valid extensions could define to get data during various stages of the ETL process.

The idea is to use the registry to dispatch the following events to extensions:
- New `ledger data` extracted
- Initial `ledger data` extracted
- Initial ledger loader sent more `objects` for the ledger

Extensions can define one or more valid hook to participate in the load process of ETL. 

The included tests check all the added safety nets (concepts) as well as enforce the correct order of hook execution.

### CODE REVIEW

File: tests/unit/etlng/RegistryTests.cpp [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+#include <xrpl/protocol/TxMeta.h>
+
+#include <cstdint>
+#include <string>
+#include <utility>
+#include <vector>
+
+using namespace etlng::impl;
+
+namespace staticTests {
```
Comment (kuznetsss) [inline tests/unit/etlng/RegistryTests.cpp]: We usually use snake_case for namespaces.

File: tests/unit/etlng/RegistryTests.cpp [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+    using spec = etlng::model::Spec<
+        ripple::TxType::ttNFTOKEN_CREATE_OFFER,
+        ripple::TxType::ttNFTOKEN_CANCEL_OFFER,
+        ripple::TxType::ttNFTOKEN_ACCEPT_OFFER>;
+    MOCK_METHOD(void, onInitialTransaction, (uint32_t, etlng::model::Transaction const&), (const));
+};
+
+}  // namespace
+
+namespace util {
```
Comment (kuznetsss) [inline tests/unit/etlng/RegistryTests.cpp]: It is better to create a test structure instead.

File: tests/unit/etlng/RegistryTests.cpp [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+{
+    auto transactions = std::vector{
+        util::CreateTransaction(ripple::TxType::ttNFTOKEN_BURN),
+        util::CreateTransaction(ripple::TxType::ttNFTOKEN_BURN),
+        util::CreateTransaction(ripple::TxType::ttNFTOKEN_CREATE_OFFER),
+    };
+
+    auto ext = MockExtLedgerData{};
+
+    EXPECT_CALL(ext, onLedgerData(testing::_)).Times(1);  // 1 batch (dispatch call)
```
Comment (kuznetsss) [inline tests/unit/etlng/RegistryTests.cpp]: `Times(1)` is by default.

File: tests/unit/etlng/RegistryTests.cpp [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+    });
+}
+
+TEST(RegistryTest, InitialObjectsCorrectOrderOfHookCalls)
+{
+    uint32_t order = 0;
+    auto extObjs = MockExtInitialObjects{};
+    auto extObj = MockExtInitialObject{};
+
+    EXPECT_CALL(extObjs, onInitialObjects).Times(1).WillRepeatedly([&](auto, auto const&, auto) {
```
Comment (kuznetsss) [inline tests/unit/etlng/RegistryTests.cpp]:
```suggestion
    EXPECT_CALL(extObjs, onInitialObjects).WillOnce([&](auto, auto const&, auto) {
```

File: tests/unit/etlng/RegistryTests.cpp [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        .successors = {},
+        .header = header,
+        .rawHeader = {},
+        .seq = SEQ
+    });
+}
+
+TEST(RegistryTest, InitialObjectsCorrectOrderOfHookCalls)
+{
+    uint32_t order = 0;
```
Comment (kuznetsss) [inline tests/unit/etlng/RegistryTests.cpp]: There is `testing::Sequence` to check mock calls in sequence.

File: tests/unit/etlng/RegistryTests.cpp (hunk 1 above)
Comment (godexsoft) [reply tests/unit/etlng/RegistryTests.cpp]: We don't actually have any snake_case namespaces. i'll rename to 'compiletime::checks' i guess.

**Review by kuznetsss — DISMISSED** [review]
👍

File: src/etlng/Models.hpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+#include <cstdint>
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace etlng::model {
+namespace impl {
+
+static consteval auto
+checkNoDuplicates(auto&&... types)
```
Comment (cindyyan317) [inline src/etlng/Models.hpp]: This method seems can be used other place. Can we move it to util. Concepts.hpp?

Comment (godexsoft) [reply src/etlng/Models.hpp:44]: I'll move it if there is going to be another place to use it in the future 👍

**Review by cindyyan317 — DISMISSED** [review]

**Review by cindyyan317 — APPROVED** [review]
Thanks adding it to util 👍

**Pull request merged**